lib/shared/sdk/usbboot/index.js

```diff
@@ -228,23 +247,97 @@ class USBBootAdapter extends EventEmitter {
     /* eslint-enable lodash/prefer-lodash-method */
 
       // This is the only way we can unique identify devices
-      device.device = `${device.busNumber}:${device.deviceAddress}`
-
-      device.displayName = 'Initializing device'
-      device.description = 'Compute Module'
-      device.raw = device.device
-      device.size = null
-      device.mountpoints = []
-      device.protected = false
-      device.system = false
-      device.disabled = true
-      device.icon = 'loading'
-      device.vendor = usbIdToString(device.deviceDescriptor.idVendor)
-      device.product = usbIdToString(device.deviceDescriptor.idProduct)
-      device.adaptor = exports.name
+      device.raw = `${device.busNumber}:${device.deviceAddress}`
+
+      const result = {
+        device: device.raw,
+        raw: device.raw,
+        displayName: 'Initializing device',
+        description: 'Compute Module',
+        size: null,
+        mountpoints: [],
+        protected: false,
+        system: false,
+        disabled: true,
+        icon: 'loading',
+        vendor: usbIdToString(device.deviceDescriptor.idVendor),
+        product: usbIdToString(device.deviceDescriptor.idProduct),
+        adaptor: exports.name
+      }
+
+      if (_.isNil(this.progress[result.raw])) {
+        // TODO: Emit an error event if this fails
+        this.prepare(device, {
+          readFile: options.readFile
+        })
+      }
+
+      result.progress = this.progress[result.raw]
+      return result
+
+    // See http://bluebirdjs.com/docs/api/promise.map.html
+    }, {
+      concurrency: 5
+    }).catch((error) => {
+      callback(error)
+    }).then((devices) => {
+      this.emit('devices', devices)
+      callback(null, devices)
+    })
+
+    return this
+  }
+
+  /**
+   * @summary Prepare a usbboot device
+   * @function
+   * @private
+   *
+   * @param {Object} device - node-usb device
+   * @param {Object} options - options
+   * @param {Function} options.readFile - read file function
+   * @returns {Promise}
+   *
+   * @example
+   * const fs = Bluebird.promisifyAll(require('fs'))
+   * const usb = require('usb')
+   * const device = usb.findByIds(0x0a5c, 0x2763)
+   *
+   * adapter.prepare(device, (name) => {
+   *   return fs.readFileAsync(name)
+   * }).then(() => {
+   *   console.log('Done!')
+   * })
+   */
+  prepare (device, options) {
+    /**
+     * @summary Set device progress
+     * @function
+     * @private
+     *
+     * @param {Number} percentage - percentage
+     *
+     * @example
+     * setProgress(90)
+     */
+    const setProgress = (percentage) => {
+      debug(`%c[${device.raw}] -> ${Math.floor(percentage)}%%`, 'color:red;')
+      this.progress[device.raw] = percentage
```

Comment:
Should the progress also be a property on the device object as `device.progress`?

Reply:
This is the internal progress cache is used as the USB device changes. There is another progress property that belongs to the drive.

Reply:
Oh, I didn't see the `result.progress = this.progress[result.raw]` at first, because it used to be `device`. Nevermind then.